### PR TITLE
Predict, save, and load satellite peaks

### LINKED ISSUE #22420: Changes to PredictFractionalPeaks
Change to one input hOffset, kOffset, and lOffset and add new input parameter for order of satellites.
Add wavelength filter like PredictPeaks.
Have peak object store both integer hkl and order of each fractional peak.
Offsets should be stored in PeaksWorkspace or peaks.

### PR DESCRIPTION
**Pull request by VickieLynch**
**Description of work.**
Creates new algorithm, PredictSatellitePeaks, which has different input from PredictFractionalPeaks which we hope to depricate. Saves satellites by integer HKL peaks and order in SaveIsawPeeaks and reads them in LoadIsawPeaks.  Offset vectors are saved in header of the file.

**To test:**
```python
peaks = LoadIsawPeaks("TOPAZ_3007.peaks")
LoadIsawUB(peaks,"TOPAZ_3007.mat")
IndexPeaks(peaks)
PredictSatellitePeaks(Peaks='peaks', SatellitePeaks='sat', ModVector1='0.4904,0.441,-0.3849',MaxOrder=1)
SaveIsawPeaks(Filename='/home/vel/workspace/sat.peaks', InputWorkspace='sat')
LoadIsawPeaks(Filename='/home/vel/workspace/sat.peaks', OutputWorkspace='sat2')
SaveIsawPeaks(Filename='/home/vel/workspace/sat2.peaks', InputWorkspace='sat2')
```

Fixes #22420 

Does this update require release notes?
- [ X] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
I have a few concerns from code review.

I also realised after reviewing most of this code that most of this is actually copied from PredictSatellitePeaks and that it is only the input that is significantly different.

Instead of duplicating the code I think it would be better to have a common class that is shared between `PredictFractionalPeaks` and this.

File: Framework/Crystal/src/LoadIsawPeaks.cpp [hunk 1]
```diff
@@ -295,9 +318,25 @@ DataObjects::Peak LoadIsawPeaks::readPeak(PeaksWorkspace_sptr outWS,
 
   seqNum = std::stoi(getWord(in, false));
 
-  h = std::stod(getWord(in, false), nullptr);
-  k = std::stod(getWord(in, false), nullptr);
-  l = std::stod(getWord(in, false), nullptr);
+  h = qSign * std::stod(getWord(in, false), nullptr);
+  k = qSign * std::stod(getWord(in, false), nullptr);
+  l = qSign * std::stod(getWord(in, false), nullptr);
+  V3D mod = V3D(0, 0, 0);
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/LoadIsawPeaks.cpp]: mod is used later in setModStru so it cannot be moved inside

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                                                          true);
+  OutPeaks->mutableRun().addProperty<std::vector<double>>("Offset2", offsets2,
+                                                          true);
+  OutPeaks->mutableRun().addProperty<std::vector<double>>("Offset3", offsets3,
+                                                          true);
+
+  V3D hkl;
+  int peakNum = 0;
+  const auto NPeaks = Peaks->getNumberPeaks();
+  Kernel::Matrix<double> Gon;
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: The PeaksWorkspace can have peaks with different goniometers so I assume we should not just use one per workspace.

[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: Ah sorry, I see later on that we're not just using the `Gon` from the first peak, but from every peak.

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 3]
```diff
@@ -0,0 +1,389 @@
...
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+
+  setPropertySettings(
+      "MaxDSpacing",
+      Kernel::make_unique<Kernel::EnabledWhenProperty>(
+          string("IncludeAllPeaksInRange"), Kernel::IS_EQUAL_TO, "1"));
+}
+
+/// Run the algorithm
+void PredictSatellitePeaks::exec() {
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: Made new function but it has lots of arguments

**Comment by VickieLynch** [comment]
I have all your changes made with the exception of sharing duplicate coding with PredictFractionalPeaks.  We have more changes already in progress, so I would prefer to wait until this project is finished to share coding or deprecate this algorithm.

**Comment by samueljackson92** [comment]
Ok I'll re-review as soon as I can. I've got release testing to do this afternoon and probably some of tomorrow.

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+  const auto NPeaks = Peaks->getNumberPeaks();
+  Kernel::Matrix<double> goniometer;
+  goniometer.identityMatrix();
+
+  const double lambdaMin = getProperty("WavelengthMin");
+  const double lambdaMax = getProperty("WavelengthMax");
+  IPeak &peak0 = Peaks->getPeak(0);
+
+  std::vector<V3D> possibleHKLs;
+  if (includePeaksInRange) {
```
Comment (samueljackson92) [inline Framework/Crystal/src/PredictSatellitePeaks.cpp]: Looking closer at this, I think you can simplify the way this works. Fundamentally I think we have two cases:

 - Predict peaks using the `HKLGenerator`
 - Predict peaks using the peaks in the peaks workspace.

Now that we've refactored the core part of the algorithm into the `predictOffsets` function I think it might be an idea to have two separate functions for the cases above. There might be a little code duplication but it would be much clearer to read what is happening.

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+    N = possibleHKLs.size();
+    N = max<size_t>(100, N);
+  }
+  auto RunNumber = peak0.getRunNumber();
+  auto &UB = lattice.getUB();
+  goniometer = peak0.getGoniometerMatrix();
+  Progress prog(this, 0.0, 1.0, N);
+  vector<vector<int>> AlreadyDonePeaks;
+  bool done = false;
+  DblMatrix orientedUB = goniometer * UB;
```
Comment (samueljackson92) [inline Framework/Crystal/src/PredictSatellitePeaks.cpp]: Could use `auto`?

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Some more comments. I haven't started on the other algorithms yet. As this PR is getting rather large I've only gone through the first algorithm. I'll post these comments now, and continue reviewing.

It might be an idea to think about whether any of the comments apply to the other algorithms in this PR.

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+
+        peak->setHKL(hkl1);
+        peak->setIntHKL(hkl);
+        peak->setPeakNumber(seqNum);
+        seqNum++;
+        peak->setRunNumber(RunNumber);
+        peak->setIntMNP(V3D(order, 0, 0));
+        OutPeaks->addPeak(*peak);
+      }
+    } catch (std::runtime_error &) {
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: Could not get an error so removed

File: Framework/Crystal/src/PredictSatellitePeaks.cpp [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+    if (Qs[2] <= 0)
+      continue;
+
+    try {
+      auto peak(Peaks->createPeak(Qs, 1));
+
+      peak->setGoniometerMatrix(goniometer);
+
+      if (peak->findDetector(tracer)) {
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: Without this line, all the detector IDs are -1 so this is the only ray tracing done.  I did fllip the condition.

[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: Uses 2nd peak constructor which does not set detector ID

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
More comments. I think I've had a first pass at everything new now.

File: Framework/Crystal/src/PredictSatellitePeaks.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/PredictSatellitePeaks.cpp]: But it does set QLab which will find the detector position via ray tracer on line `601` if a detector distance is not passed.

**Comment by VickieLynch** [comment]
See changes to PeakCol: Do you know of anyone who changes the RunNumber in the peaks table manually.  We are having trouble with our 5 digit RunNumbers having a comma inserted by locale and then when reset from the table workspace having what is after the comma truncated.  I tried earlier to remove that comma and something was changed in Mantid where the other read was used and this error came back.

**Comment by samueljackson92** [comment]
@VickieLynch I'm not sure. I've seen this error before too, but I agree that thought this had been fixed. Can you reproduce it? If so can you make a bug report?

**Comment by VickieLynch** [comment]
RunNumber bug is fixed in #23161 so change in PeaksColumn will be in master soon.

File: Framework/MDAlgorithms/src/CentroidPeaksMD2.cpp [hunk 8]
```diff
@@ -115,6 +115,7 @@ void CentroidPeaksMD2::integrate(typename MDEventWorkspace<MDE, nd>::sptr ws) {
           centroid[d] /= static_cast<coord_t>(signal);
 
         V3D vecCentroid(centroid[0], centroid[1], centroid[2]);
+        p.setBinCount(static_cast<double>(signal));
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/CentroidPeaksMD2.cpp]: In pull request #23202

File: docs/source/algorithms/IndexPeakswithSatellites-v1.rst [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+.. properties::
+
+Description
+-----------
+
+This Algorithm indexes a PeaksWorkspace with peaks occurring at specific
+fractional offsets from h,k,or l values.
+The order can be 1 for +/-1 or 2 for +/-1 and +/-2, etc.
+
+The input PeaksWorkspace must contain an orientation matrix.
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply docs/source/algorithms/IndexPeakswithSatellites-v1.rst]: Many of the Crystal algorithms require this so this is a good idea

File: Framework/DataObjects/inc/MantidDataObjects/Peak.h [hunk 10]
```diff
@@ -152,6 +154,8 @@ class DLLExport Peak : public Geometry::IPeak {
   void setCol(int m_col);
   void setPeakNumber(int m_peakNumber) override;
   int getPeakNumber() const override;
+  void setIntMNP(const Mantid::Kernel::V3D m_modStru) override;
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/DataObjects/inc/MantidDataObjects/Peak.h]: MNP is like HKL but for the modulations so this is the preferred name

File: Framework/Crystal/src/IndexPeakswithSatellites.cpp [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+        } else {
+          for (int m = -maxOrder; m <= maxOrder; m++)
+            for (int n = -maxOrder; n <= maxOrder; n++)
+              for (int p = -maxOrder; p <= maxOrder; p++) {
+                if (m == 0 && n == 0 && p == 0)
+                  continue; // exclude 0,0,0
+                V3D hkl1(hkl);
+                hkl1[0] -= m * offsets1[0] + n * offsets2[0] + p * offsets3[0];
+                hkl1[1] -= m * offsets1[1] + n * offsets2[1] + p * offsets3[1];
+                hkl1[2] -= m * offsets1[2] + n * offsets2[2] + p * offsets3[2];
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/IndexPeakswithSatellites.cpp]: Tried this:  no match for ‘operator*’ (operand types are ‘Mantid::Kernel::V3D’ and ‘Mantid::Kernel::DblMatrix

[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/IndexPeakswithSatellites.cpp]: But `DblMatrix` has a `operator*` that I think does what you want? It also looks like `DblMatrix` has support for the matrix operator and a `std::vector` as well...

**Comment by VickieLynch** [comment]
Shiyun has found a problem with PredictSatellites output that I need to fix before this is merged.

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Sorry Vickie, I know you're keen to get this in, but I still have a few issues with this stuff. 

If it's in a working state and your users are desperate for it then perhaps we could start doing some of this stuff on separate refactoring PRs?

My only concern is that things go in and then the refactoring never gets done.

**Comment by VickieLynch** [comment]
````python
LoadIsawPeaks(Filename='24281_predict.peaks', OutputWorkspace='peaks')
SaveIsawPeaks(Filename='24281_predict.peaks2', InputWorkspace='peaks')
FindUBUsingIndexedPeaks(PeaksWorkspace='peaks', Tolerance=0.12)
IndexPeaks(PeaksWorkspace='peaks', Tolerance=0.12)
IndexPeaksWithSatellites(PeaksWorkspace='peaks', Tolerance=0.12, ModVector1='-0.49088,-0.442061,0.384217', MaxOrder=2)
````

**Comment by VickieLynch** [comment]
https://www.dropbox.com/preview/24281_predict.peaks?role=work

Try this input file with above script for testing

**Comment by VickieLynch** [comment]
[24281_predict.peaks.txt](https://github.com/mantidproject/mantid/files/2311028/24281_predict.peaks.txt)

**Comment by VickieLynch** [comment]
Tested satellite integration with:
````python
LoadIsawPeaks(Filename='24281_predict.peaks', OutputWorkspace='peaks')
FindUBUsingIndexedPeaks(PeaksWorkspace='peaks', Tolerance=0.12)
Load(Filename='/SNS/TOPAZ/IPTS-18749/data/TOPAZ_24281_event.nxs', OutputWorkspace='TOPAZ_24281_event')
IntegrateEllipsoidsWithSatellites(InputWorkspace='TOPAZ_24281_event', PeaksWorkspace='peaks', SpecifySize=True, OutputWorkspace='peaks')

````

File: Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp [hunk 12]
```diff
@@ -87,6 +92,10 @@ void export_IPeak() {
       .def("getL", &IPeak::getL, arg("self"), "Get the L index of the peak")
       .def("getHKL", &IPeak::getHKL, arg("self"),
            "Get HKL as a :class:`~mantid.kernel.V3D` object")
+      .def("getIntHKL", &IPeak::getIntHKL, arg("self"),
+           "Get HKL as a :class:`~mantid.kernel.V3D` object")
+      .def("setIntHKL", &IPeak::setIntHKL, (arg("self"), arg("hkl")),
+           "Set the integer HKL for this peak")
```
Comment (samueljackson92) [inline Framework/PythonInterface/mantid/api/src/Exports/IPeak.cpp]: Also needs python unit tests for these as well.

File: Framework/Crystal/src/IndexPeaks.cpp [hunk 13]
```diff
@@ -168,29 +207,248 @@ void IndexPeaks::exec() {
...
+                        peaks[i].setIntMNP(V3D(m, n, p));
+                        suc_indexed = true;
+                        sate_indexed++;
+                        h_error = fabs(round(hkl1[0]) - hkl1[0]);
+                        k_error = fabs(round(hkl1[1]) - hkl1[1]);
+                        l_error = fabs(round(hkl1[2]) - hkl1[2]);
+                        sate_error += h_error + k_error + l_error;
+                      }
+                    }
+              }
```
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/IndexPeaks.cpp]: In fact, you might be able to get away with just using `HKLGenerator` itself here, maybe with a new constructor.

**Review by samueljackson92 — CHANGES_REQUESTED** [review]
Ok, so I've had a quick run through with the WISH data I have and the algorithms seem to work ok. Or at least they give me answers that I can believe. 🎉 

Sorry it has taken me a while to review, but there's a fair amount in here to go through.

I've added a bunch more review comments because I'm still not happy with the quality of a fair amount of the code in here and I need to record that somewhere. But I am aware that your users are after these features. I feel that these comments need to be addressed, but am willing to be flexible letting things go in if people really need them. We should discuss a plan for how we tackle this.

Additionally I spotted the following minor things when testing:
 - There doesn't seem to be error checking on how many numbers I pass when supplying a modulation vector to `PredictSatellitePeaks`. i.e. I would expect I need to pass exactly 3 and it should moan if I don't.
 - `PredictSatellitePeaks` has no progress bar to indicate how long it will run for.

I also think there are several other pieces of work that need to be done before we completely done with modulated structures:
 - Cannot overlay or show the 3+d indexing in the IV and slice viewer.
 - I think this additional data should be shown in the table view for peaks workspace, both in general and in the slice viewer.
 - We're missing python exports for the new methods on unit cell etc.

But please don't address those in this PR as it's already so enormous and therefore takes significant effort to review. If you need to add more new functionality can you please make a new branch and a new PR based off of this one? (minor changes/refactoring should however still be done here).

File: Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp [hunk 14]
```diff
@@ -82,31 +118,151 @@ void FindUBUsingIndexedPeaks::exec() {
...
+    for (Peak peak : peaks) {
+      q_vectors.push_back(peak.getQSampleFrame());
     }
-    double tolerance = this->getProperty("Tolerance");
+
     int num_indexed = IndexingUtils::NumberIndexed(UB, q_vectors, tolerance);
+    int sate_indexed = 0;
+    double satetolerance = getProperty("ToleranceForSatellite");
+
+    if (ModDim > 0) {
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp]: ../Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp:125:22: error: continue statement not within a loop
     if (ModDim == 0) continue;

File: Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp [hunk 15]
```diff
@@ -82,31 +118,151 @@ void FindUBUsingIndexedPeaks::exec() {
...
+                errorHKL2 += errhkl;
+              }
+            } else if (run_mnp_vectors[i][2] != 0) {
+              fhkl -= run_lattice.getModVec(3) * run_mnp_vectors[i][2];
+              if (IndexingUtils::ValidIndex(fhkl, satetolerance)) {
+                sate_indexed++;
+                V3D errhkl = fhkl - run_hkl_vectors[i];
+                errhkl = V3D(fabs(errhkl[0]), fabs(errhkl[1]), fabs(errhkl[2]));
+                errorHKL3 += errhkl;
+              }
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp]: No, I looked at this using PredictSatellitePeaks for data with 2 vectors.  Glad you asked because the CrossTerms were not done correctly so I fixed so it indexes predicted peaks correctly.  
````python
LoadIsawPeaks(Filename='/SNS/TOPAZ/IPTS-10003/shared/K2V3O8_satellite_5c_transformed/8071_combined.integrate', OutputWorkspace='peaks')
LoadIsawUB(InputWorkspace='peaks', Filename='/SNS/TOPAZ/IPTS-10003/shared/K2V3O8_satellite_5c_transformed/8071_combined.mat')
PredictSatellitePeaks(Peaks='peaks', SatellitePeaks='cross', MaxOrder=2, GetModVectorsFromUB=True, CrossTerms=True, IncludeAllPeaksInRange=True, WavelengthMin=1, WavelengthMax=2, MinDSpacing=0.5, MaxDSpacing=3)
FindUBUsingIndexedPeaks(PeaksWorkspace='cross', ToleranceForSatellite=0.050000000000000003)
````

File: Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp]: My mistake you're not in a loop. This is a little more complicated to restructure.

File: Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp (hunk 15 above)
[the remark this answers is not in the document]
Comment (samueljackson92) [reply Framework/Crystal/src/FindUBUsingIndexedPeaks.cpp]: > Glad you asked because the satellites were not done for CrossTerms=True so I fixed so it indexes predicted peaks correctly.

So did you also add a unit test case that covers this?

File: docs/source/concepts/ModulatedStructure.rst [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+.. math:: ModHKL = \text{UB}^{- 1} \times ModUB
+
+New functions setModUB is added to OrientedLattice class, and setModHKL,
+getModHKL, setModerr1, setModerr2, setModerr3, setModVec1, setModVec2,
+setModVec3, getModerr(i), getModVec(i), getdh(i), getdherr(i), getdk(i),
+getdkerr(i), getdl(i), getdlerr(i) are added to UnitCell class. Value
+for ModHKL in UnitCell is set when the function setModUB is used in
+OrientedLattice.
+
+A python script is created for step 4, which provides a visual aid for
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply docs/source/concepts/ModulatedStructure.rst]: The python script is the one where you tried clustering. It is not quite ready for inclusion in Mantid since it has no user interface. I will reword.

File: Framework/Crystal/src/IndexPeaks.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/Crystal/src/IndexPeaks.cpp]: Reduced lines of code with V3D call

File: Framework/MDAlgorithms/test/IntegrateEllipsoidsWithSatellitesTest.h [hunk 17]
```diff
@@ -0,0 +1,465 @@
+// Mantid Repository : https://github.com/mantidproject/mantid
+//
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/MDAlgorithms/test/IntegrateEllipsoidsWithSatellitesTest.h]: Added satellite peaks to test

File: Framework/MDAlgorithms/src/Integrate3DEvents.cpp [hunk 18]
```diff
@@ -56,6 +57,51 @@ Integrate3DEvents::Integrate3DEvents(
...
+        getHklMnpKey(boost::math::iround<double>(hkl_list[it][0]),
+                     boost::math::iround<double>(hkl_list[it][1]),
+                     boost::math::iround<double>(hkl_list[it][2]),
+                     boost::math::iround<double>(mnp_list[it][0]),
+                     boost::math::iround<double>(mnp_list[it][1]),
+                     boost::math::iround<double>(mnp_list[it][2]));
+    if (hklmnp_key != 0) // only save if hkl != (0,0,0)
+      m_peak_qs[hklmnp_key] = peak_q_list[it].second;
+  }
+}
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/Integrate3DEvents.cpp]: unit test added

File: Framework/MDAlgorithms/src/IntegrateEllipsoidsWithSatellites.cpp [hunk 19]
```diff
@@ -0,0 +1,753 @@
...
+#include "MantidKernel/Statistics.h"
+#include "MantidMDAlgorithms/Integrate3DEvents.h"
+#include "MantidMDAlgorithms/MDTransfFactory.h"
+#include "MantidMDAlgorithms/MDTransfQ3D.h"
+#include "MantidMDAlgorithms/UnitsConversionHelper.h"
+
+#include <boost/math/special_functions/round.hpp>
+#include <cmath>
+
+using namespace Mantid::API;
```
[the remark this answers is not in the document]
Comment (VickieLynch) [reply Framework/MDAlgorithms/src/IntegrateEllipsoidsWithSatellites.cpp]: The idea was to have one algorithm when we trusted the new one was working correctly. Would renaming the new one to be the standard ellipsoid integration satisfy what you want here?

File: Framework/Crystal/src/IndexPeaks.cpp [hunk 20]
```diff
@@ -44,11 +55,27 @@ void IndexPeaks::init() {
...
+
+  this->declareProperty(
+      make_unique<PropertyWithValue<double>>("MainError", 0.0,
+                                             Direction::Output),
+      "Gets set with the average HKL indexing error of Main Peaks.");
+
+  this->declareProperty(
+      make_unique<PropertyWithValue<double>>("SatelliteError", 0.0,
+                                             Direction::Output),
+      "Gets set with the average HKL indexing error of Satellite Peaks.");
```
Comment (samueljackson92) [inline Framework/Crystal/src/IndexPeaks.cpp]: There's no testing for these new parameters? How do we know if it works? How do we know when it breaks?

File: Framework/Crystal/test/IndexPeaksWithSatellitesTest.h [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+#include "MantidKernel/System.h"
+#include "MantidKernel/Timer.h"
+#include <cxxtest/TestSuite.h>
+
+using namespace Mantid::Crystal;
+using namespace Mantid::API;
+using namespace Mantid::DataObjects;
+using namespace Mantid::Kernel;
+
+class IndexPeaksWithSatellitesTest : public CxxTest::TestSuite {
```
Comment (samueljackson92) [inline Framework/Crystal/test/IndexPeaksWithSatellitesTest.h]: I'm still not satisfied with the level of testing here. Have we really tested a decent subset of the edge cases? Some cases to think about:
 - Different numbers/ combinations of mod vectors?
 - With/without *RoundHKLs*
 - With/without *CommonUBForAll*
 - With/without *GetVectorsFromUB*
 - With/without *CrossTerms*
 - Testing the output properties actually return the correct values.

Please try to add simple test cases for each of these, and not test everything in one large method.

File: Framework/Crystal/inc/MantidCrystal/LoadIsawPeaks.h [hunk 22]
```diff
@@ -87,6 +93,11 @@ class DLLExport LoadIsawPeaks
   boost::shared_ptr<const Geometry::IComponent> getCachedBankByName(
       std::string bankname,
       const boost::shared_ptr<const Geometry::Instrument> &inst);
+
+  /// Read offsets and add as property
+  void getOffsets(std::ifstream &in, DataObjects::PeaksWorkspace_sptr Peaks,
+                  std::vector<double> &m_offset, std::string &label,
+                  double &qSign);
```
Comment (samueljackson92) [inline Framework/Crystal/inc/MantidCrystal/LoadIsawPeaks.h]: Can we add a test to LoadIsawPeaksTest for modulated structures?

File: Framework/Crystal/inc/MantidCrystal/LoadIsawUB.h [hunk 23]
```diff
@@ -45,6 +46,8 @@ class DLLExport LoadIsawUB : public API::Algorithm {
   void init() override;
   /// Run the algorithm
   void exec() override;
+  /// load the modulation UB
+  void readModulatedUB(std::ifstream &in, Kernel::DblMatrix &ub);
```
Comment (samueljackson92) [inline Framework/Crystal/inc/MantidCrystal/LoadIsawUB.h]: Can we add a test to LoadIsawUB for modulated UBs?

**Review by samueljackson92 — COMMENTED** [review]
I understand your need to get this in because of user pressure and that getting user validation on this work is also useful & important. However, I still have big concerns about the technical debt that will be introduced by this PR as it is and I am therefore reluctant to approve it.

So I suggest a compromise: Lets only focus on getting the test coverage for the new work up to the best level possible. If we have good test coverage, then it can be easily & safely refactored in another ticket later in a follow up issue. 

Therefore In this review I have only made comments suggesting improvements to the tests. To summarise what we need to do:

 - [ ] Add testing for modulated peaks in `LoadIsawPeaks`
 - [ ] Add testing for modulated peaks in `SaveIsawPeaks`
 - [ ] Add testing for modulated peaks in `LoadNexusProcessed`
 - [ ] Add testing for modulated peaks in `SaveNexusProcessed`
 - [ ] Add testing for modulated UBs in `LoadIsawUB`
 - [ ] Add testing for modulated UBs in `SaveIsawUB`
 - [ ] Increase the level of coverage for `IndexPeaksWithSatellitesTest` for untested properties
 - [ ] Increase the level of coverage for `IndexPeaksTest` for new functionality
 - [ ] Increase the level of coverage for `PredictSatellitePeaks` for untested properties
 - [ ] Add a performance test for `PredictSatellitePeaks` 
 - [ ] Add test coverage for `OrientedLattice` for new functionality.
 - [ ] Add test coverage for `IndexingUtils` for new functionality.
 - [ ] Add test coverage for `Peak` for new functionality.
 - [ ] Add test coverage for `UnitCell` for new functionality.
 - [ ] Add test coverage for `V3D` for new functionality.

Some of these (e.g. `V3D`) should require only a small amount of work to test. If we have this then we have some confidence that everything in this PR works as we expect and we can have some confidence about not breaking that functionality when we refactor.

File: docs/source/concepts/ModulatedStructure.rst [hunk 24]
```diff
@@ -0,0 +1,412 @@
...
+is returned.  This method is used to optimize the UB matrix and ModUB matrix
+once an initial indexing has been found. Other than ModUB and
+the list of mnp vectors as additional arguments for the function, a
+const int ModDim is also added to describe the modulation dimension of
+the indexed peaks list. In the case of modulation dimension equals
+three:
+
+.. math::
+
+   {\begin{pmatrix}
```
Comment (rosswhitfield) [inline docs/source/concepts/ModulatedStructure.rst]: This doesn't compile for me, need to remove the outermost `{}`

File: docs/source/concepts/ModulatedStructure.rst [hunk 25]
```diff
@@ -0,0 +1,412 @@
...
+   \begin{matrix}
+   m_{i} \\
+   n_{i} \\
+   p_{i} \\
+   \end{matrix} \\
+   \end{pmatrix}}
+
+.. math::
+
+   {\begin{pmatrix}
```
Comment (rosswhitfield) [inline docs/source/concepts/ModulatedStructure.rst]: Same here, remove outermost `{}`

File: Framework/Geometry/inc/MantidGeometry/Crystal/UnitCell.h [hunk 26]
```diff
@@ -20,26 +20,46 @@
...
+
+ References:
+ - International Tables for Crystallography (2006). Vol. B, ch. 1.1, pp. 2-9
+ - W. R. Busing and H. A. Levy, Angle calculations for 3- and 4-circle X-ray
+ and neutron diffractometers - Acta Cryst. (1967). 22, 457-464
+
+
+ @author Andrei Savici, SNS, ORNL
+ @date 2011-03-23
+ Copyright &copy; 2007-8 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
```
Comment (rosswhitfield) [inline Framework/Geometry/inc/MantidGeometry/Crystal/UnitCell.h]: You appear to have reverted the license style changes here, I suggest look at this file in `master` to make it consistent.

File: Framework/Crystal/inc/MantidCrystal/IndexPeaksWithSatellites.h [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+  };
+  /// Summary of algorithms purpose
+  const std::string summary() const override {
+    return "Index the peaks using the UB from the sample.";
+  }
+
+  /// Algorithm's version for identification
+  int version() const override { return 1; }
+  const std::vector<std::string> seeAlso() const override {
+    return {"IndexPeaksWithSatellites"};
```
Comment (rosswhitfield) [inline Framework/Crystal/inc/MantidCrystal/IndexPeaksWithSatellites.h]: This is referencing itself for `seeAlso`?

File: Framework/MDAlgorithms/inc/MantidMDAlgorithms/Integrate3DEvents.h [hunk 28]
```diff
@@ -37,17 +37,40 @@ struct IntegrationParameters {
...
+ find the net integrated counts, using ellipsoids with axis lengths
+ determined from the standard deviations in the directions of the
+ principal axes.
 
-    @author Dennis Mikkelson
-    @date   2012-12-19
+ @author Dennis Mikkelson
+ @date   2012-12-19
+
+ Copyright © 2012 ORNL, STFC Rutherford Appleton Laboratories
```
Comment (rosswhitfield) [inline Framework/MDAlgorithms/inc/MantidMDAlgorithms/Integrate3DEvents.h]: Also here, the license from here down should no longer in the file.

**Review by rosswhitfield — CHANGES_REQUESTED** [review]
Some initial comments, probably more to come.

You should also remove the commented out code that Sam pointed to.

**Comment by rosswhitfield** [comment]
There are unit tests and a docs page for the algorithm `IntegrateEllipsoidsWithSatellites` but the algorithm itself doesn't appear to exist. Was this accidentally removed at some point?

**Comment by rosswhitfield** [comment]
I tried your test instructions and when I load the peaks back in they all have integer hkl's. `SaveIsawPeaks` doesn't appear to save the offset vectors.

**Comment by VickieLynch** [comment]
> There are unit tests and a docs page for the algorithm `IntegrateEllipsoidsWithSatellites` but the algorithm itself doesn't appear to exist. Was this algorithm accidentally removed at some point? Or if it was deliberate then the docs and test should be removed.

Removed doc page. The two algorithms were merged, so I left the WithSatellites tests to test IntegrateEllipsoids. These two files can be merged if you prefer.

**Comment by VickieLynch** [comment]
> I tried your test instructions and when I load the peaks back in they all have integer hkl's. `SaveIsawPeaks` doesn't appear to save the offset vectors, the method `writeOffsets` is never used.

The offset vectors are recalculated when LoadIsawPeaks is loaded. We do not show the mnp indices when the peaks are in the table, but they are in memory and could be added.

**Comment by rosswhitfield** [comment]
It's fine that the test `IntegrateEllipsoidsWithSatellites` is separate but appears to not actually run? `ctest -R Satellites` only shows `CrystalTest_IndexPeaksWithSatellitesTest`.

**Comment by NickDraper** [comment]
```
There once was a branch with conflicts.
Delays to the PR this predicts.
His colleague urged
with master he merged,
and then those conflicts he could fix.
```

**Review by rosswhitfield — DISMISSED** [review]

File: Framework/Crystal/inc/MantidCrystal/IndexPeaksWithSatellites.h [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+namespace Mantid {
+namespace Crystal {
+/** IndexPeaksWithSatellites : Algorithm to use the UB saved in the sample
+   associated
+    with the specified PeaksWorkspace, to index the peaks in the workspace.
+
+    @author Shiyun Jin
+    @date   2018-08-17
+
+    Copyright &copy; 2011 ISIS Rutherford Appleton Laboratory &
```
Comment (soininen) [inline Framework/Crystal/inc/MantidCrystal/IndexPeaksWithSatellites.h]: This and the other new files should have the new-style copyright notice at the very beginning of the file. They also tell me it is 2019 already. Maybe the years could be updated on the same go?

**Review by soininen — CHANGES_REQUESTED** [review]
As the comment below suggests, the copyright notices could be updated to the latest style.

**Review by soininen — APPROVED** [review]

**Review by rosswhitfield — APPROVED** [review]
